OvmfPkg/PlatformPei/PlatformId.c

```diff
@@ -0,0 +1,660 @@
+  IN UINT32  LocatorType
+  )
+{
+  UINTN  Size;
+
+  // No validation done for non-UEFI variable locators.
+  if (LocatorType != LOCATOR_TYPE_UEFI_VARIABLE) {
+    return EFI_SUCCESS;
+  }
+
+  Size = (UINTN)(LocatorEnd - Locator);
+  // The size needs to include the GUID and a UCS2 NUL terminated string, so it
+  // has to be at least 18 bytes, and be even.
+  if ((Size < sizeof (EFI_GUID) + 2) || ((Size & 1) != 0)) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "UEFI variable locator improper size %d\n", Size));
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  if (*((CHAR16 *)LocatorEnd - 1) != 0) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "UEFI variable locator not UCS2 0-terminated\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  return EFI_SUCCESS;
+}
+
+typedef struct tdTCG_Sp800_155_PlatformId_Event3_Cursors {
+  UINT32      VendorId;
+  EFI_GUID    ReferenceManifestGuid;
+  UINT8       *PlatformManufacturerStr;
+  UINT8       *PlatformManufacturerStrEnd;
+  UINT8       *PlatformModel;
+  UINT8       *PlatformModelEnd;
+  UINT8       *PlatformVersion;
+  UINT8       *PlatformVersionEnd;
+  UINT8       *FirmwareManufacturerStr;
+  UINT8       *FirmwareManufacturerStrEnd;
+  UINT32      FirmwareManufacturerId;
+  UINT8       *FirmwareVersion;
+  UINT8       *FirmwareVersionEnd;
+  UINT32      RimLocatorType;
+  UINT8       *RimLocator;
+  UINT8       *RimLocatorEnd;
+  UINT32      PlatformCertLocatorType;
+  UINT8       *PlatformCertLocator;
+  UINT8       *PlatformCertLocatorEnd;
+} TCG_Sp800_155_PlatformId_Event3_Cursors;
+
+/**
+ * Parses an SP800155 Event3 structure from the buffer [Evt, BufferEnd).
+ *
+ * All variable length fields are represented as [Start, End) pointers and do
+ * not include the length. It is always the case that Start <= End if
+ * EFI_SUCCESS is returned.
+ *
+ * @param[in] Evt  The start of the Event3 structure.
+ * @param[in] BufferEnd  The end of the buffer.
+ * @param[out] Cursors  The parsed structure.
+ *
+ * @retval EFI_SUCCESS  The structure was parsed successfully.
+ * @retval EFI_BUFFER_TOO_SMALL  The buffer is too small for the required fields
+ *                               and declared variable lengths of buffers.
+ * @retval EFI_INVALID_PARAMETER  The structure is invalid.
+ */
+STATIC
+EFI_STATUS
+ParseSp800155 (
+  IN  UINT8                                    *Evt,
+  IN  UINT8                                    *BufferEnd,
+  OUT TCG_Sp800_155_PlatformId_Event3_Cursors  *Cursors
+  )
+{
+  EFI_STATUS  Status;
+  UINT8       *Cursor;
+  BOOLEAN     Event2;
+  BOOLEAN     Event3;
+
+  // UINT8 Signature[16]
+  Status = SafeOffset (Evt, BufferEnd, 16, &Cursor);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, DPREFIX "Signature %d\n", Status));
+    return Status;
+  }
+
+  Event2 = CompareMem (
+             Evt,
+             TCG_Sp800_155_PlatformId_Event2_SIGNATURE,
+             sizeof (TCG_Sp800_155_PlatformId_Event2_SIGNATURE) - 1
+             ) == 0;
+  Event3 = !Event2 && (CompareMem (

```

Comment:
why put `!Event2` here?

Reply:
It's a logic shortcut to avoid a longer comparison. It's like having `if (...) { Event2 = True; } else if (...) { Event3 = True; }`

Reply:
I would rather you write in this way `if (...) { Event2 = True; } else if (...) { Event3 = True; }`
It is more human readable.

Reply:
Removed for uninterpreted single file solution.